README.md

```diff
@@ -593,6 +593,61 @@ mysetting: |
 
 The possibility is endless. Try importing values from your golang app, bash script, jsonnet, or anything!
 
+## Hooks
+
+Hook is an per-release extension point of helmfile that is composed of:
+
+- `events`
+- `command`
+- `args`
+
+helmfile triggers various `events` while it is running.
+One `events` are triggered, associated `hooks` are executed, by running the `command` with `args`. 
+
+Currently supported `events` are:
+
+- `prepare`
+- `cleanup`
+
+Hooks associated to `prepare` events are triggered after each release in your helmfile is loaded from YAML, before executed.
+
+Hooks associated to `cleanup` events are triggered after each release is processed.
+
+The following is an example hook that just prints the contextual information provided to hook:
+
+```
+releases:
+- name: myapp
+  chart: mychart
+  # *snip*
+  hooks:
+  - events: ["prepare", "cleanup"]
```

Comment:
Is there a technical reason you didn't go with this the structure proposed in #330?:

```
 hooks:
    preChartLoad: # register as many commands as you want to this hook
    - command: {{` ./chartify -e {{ .Environment.Name }} {{ .Chart }} `}}
```

I see that you would want to register the same command for a specific hook, but that seems to be over generic (I believe it is more common to have differing calls on each event, than the same call on multiple events).

YAML anchors could come handy if you have the same command on multiple events.

The `preChartLoad` schema is verifiable at load time.

Reply:
Good catch! Indeed my intention was to make it even more generic with (hopefully) comparable complexity to the third option in [my proposal](https://github.com/roboll/helmfile/issues/295#issuecomment-419870215).

> I believe it is more common to have differing calls on each event, than the same call on multiple events

I think so too. My intention here was to provide the most DRY way to support this use-case:

> To be able to use the same script to implement multiple hooks the name of the hook should also be passed along.

See https://github.com/roboll/helmfile/issues/295#issuecomment-419051383 for more context.

> The preChartLoad schema is verifiable at load time.

Good point! I was thinking to add validation for unsupported events. That is, `events: ["foo"]` results in an error like `unsupported event to hook: foo`.

Reply:
And I wasn't sure if "events` are the only possible hook target in helmfile. Googling around, the term `hook` seems to be used for NOT only for expressing extension points associated to events. For example, something like "action hook" can be considered:

```
hooks:
- actions: ["lint"]
   command: ...
```

An event hook reacts on specific event and doesn't have ability to return data from the hook to helmfile.

An action hook, on the other hand, could be used for passing  data from the hook to helmfile, in an output format helmfile expects for the specific action. In the above example, the `lint` action hook can return a specifically formatted json to add lint results that are integrated into the original output of `helmfile lint`.

Explicitly wording hook targets as `events` prevents ambiguity and creates future possibility like this.